pkg/controller/master/master_controller.go

```diff
@@ -157,16 +182,114 @@ func (s *installStrategy) setupServiceAccountForJob(clusterNamespace string) (*k
+	logger := logging.WithMachineSet(s.logger, masterMachineSet)
+	machines, err := s.availableMasterMachines(masterMachineSet)
+	if err != nil {
+		logger.Errorf("error determining available master machines: %v", err)
+		return false
+	}
+	if masterMachineSet.Spec.Replicas == nil {
+		logger.Errorf("no replicas defined for master machineset")
+		return false
+	}
+	// Only start an install if we have the number of replicas specified in the machineset
+	if len(machines) < int(*masterMachineSet.Spec.Replicas) {
+		return false
+	}
+	// Of the available machines, determine if any has not had its control plane installed
+	// If all of them have a control plane, then no need to run this job.
+	// TODO: determine how to handle upgrades, if they require a job.
+	for _, m := range machines {
+		if m.Status.Versions == nil || len(m.Status.Versions.ControlPlane) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func hasPublicDNS(machine *capi.Machine) (bool, error) {
+	providerStatus, err := controller.AWSMachineProviderStatusFromClusterAPIMachine(machine)
+	if err != nil {
+		return false, err
+	}
+	return providerStatus.PublicDNS != nil && len(*providerStatus.PublicDNS) > 0, nil
+}
+
 func (s *installStrategy) IncludeInfraSizeInAnsibleVars() bool {
 	return false
 }
 
 func (s *installStrategy) OnInstall(succeeded bool, cluster *clustop.CombinedCluster, masterMachineSet *capi.MachineSet, job *batchv1.Job) {
+	if succeeded {
+		machineKeys := ansible.GetJobMachineKeys(job)
+		err := s.setMachineControlPlaneVersion(machineKeys)
+		if err != nil {
+			logging.WithMachineSet(s.logger, masterMachineSet).Errorf("error updating machine installed control plane version: %v", err)
+			return
+		}
+	}
 	cluster.ClusterDeploymentStatus.ControlPlaneInstalled = succeeded
 	cluster.ClusterDeploymentStatus.ControlPlaneInstalledJobClusterGeneration = cluster.Generation
 	cluster.ClusterDeploymentStatus.ControlPlaneInstalledJobMachineSetGeneration = masterMachineSet.GetGeneration()
 }
 
+func (s *installStrategy) setMachineControlPlaneVersion(machineKeys []string) error {
+	errs := []error{}
+	for _, machineKey := range machineKeys {
+		namespace, name, err := cache.SplitMetaNamespaceKey(machineKey)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error splitting machine key (%s): %v", machineKey, err))
+		}
+		machine, err := s.machineLister.Machines(namespace).Get(name)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("error retrieving machine %s to update the control plane version", machineKey))
+		}
+		updatedMachine := machine.DeepCopy()
+		if updatedMachine.Status.Versions == nil {
+			updatedMachine.Status.Versions = &capi.MachineVersionInfo{}
+		}
+		updatedMachine.Status.Versions.ControlPlane = machine.Spec.Versions.ControlPlane
+		updatedMachine.Status.LastUpdated = metav1.Now()
+		_, err = s.capiClient.ClusterV1alpha1().Machines(machine.Namespace).UpdateStatus(updatedMachine)
```

Comment:
I am still a bit squeamish about having a controller update two different resources in a single sync. It is fine for now, but I'd like us to keep thinking about whether we can do better somehow.